torch/distributed/rpc/api.py

```diff
@@ -361,17 +362,31 @@ def _to_worker_info(to):
         raise ValueError("Cannot get WorkerInfo from name {}".format(to))
 
 
-def _rref_typeof_on_owner(rref):
-    return type(rref.local_value())
+def _rref_typeof_on_owner(rref, blocking=True):
+    rref_type = type(rref.local_value())
+    if blocking:
+        return rref_type
+    else:
+        # Wrap result into a completed Future. This is so that if blocking=`False`
+        # is specified, we return a future regardless of if this call is on user
+        # or owner.
+        future = Future()
+        future.set_result(rref_type)
+        return future
```

Comment:
if this returns a future, do we need to annotate this function with an `@rpc.functions.async_execution` decorator?

Reply:
I recall Future is not picklable, and we didn't plan to make Future picklable, as communicating a future to a remote process does not seem reasonable? E.g., if the future is not completed yet, do we need to update both local and remote futures when the local one is marked as completed?

Reply:
This won't return a future when called remotely, as the only remote call is from the below `_rref_typeof_on_user`, and we call that without blocking (just need to get the type in that case).

I didn't decorate it with rpc.functions.async_execution for that reason, because its always ran synchronously when called over RPC. So I guess this means picklability of Future isn't a concern?

I think this could potentially cause issues later if this function is ran over RPC with `blocking=False`, but since its private I'm assuming use cases will be like the ones below.